{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "#### Assembly Statistics\n",
    "\n",
    "\n",
    "contigs.fasta\n",
    "Total length = 6678635\n",
    "\n",
    "Number of contigs = 176\n",
    "\n",
    "N50 = 234690, n = 11\n",
    "\n",
    "scaffolds.fasta\n",
    "Total length = 6678655\n",
    "\n",
    "Number of contigs = 174\n",
    "\n",
    "N50 = 234690, n = 11\n",
    "\n",
    "### 16S Genus IDs\n",
    "*  NODE_23_length_103367_cov_10.081286:35530-37075\n",
    "\n",
    "*  Genus = Pseudomonas\n",
    "\n",
    "*  Came to conclusion based on definitive genus ID (no species ID available)\n",
    "\n",
    "* Querying the genus similarity, we found about 10 Pseudomonas species with an average of .98 ID with the matched portions\n",
    "\n",
    "#### NODE_18_length_154232_cov_9.584186:124541-126086\n",
    "\n",
    "* Genus = Pseudomonas\n",
    "\n",
    "* Near identical results to \"NODE_23\"\n",
    "\n",
    "#### NODE_5_length_327223_cov_9.241629:46285-47830\n",
    "\n",
    "* Genus = Pseudomonas\n",
    "\n",
    "* Near identical results to \"NODE_18\"\n",
    "\n",
    "#### NODE_6_length_298784_cov_8.918606:13925-15470\n",
    "\n",
    "* Genus = Pseudomonas\n",
    "\n",
    "* Near identical results to \"NODE_5\"\n",
    "\n",
    "#### NODE_4_length_332509_cov_9.114046:137011-138556\n",
    "\n",
    "* Genus = Pseudomonas\n",
    "\n",
    "* Near identical results to \"NODE_6\"\n",
    "\n",
    "#### NODE_2_length_378948_cov_10.006665:348499-350044\n",
    "\n",
    "* Genus = Pseudomonas\n",
    "\n",
    "* Near identical results to \"NODE_4\"\n",
    "\n",
    "### FINAL GENUS-LEVEL IDENTIFICATION\n",
    "Pseudomonas\n",
    "\n",
    "Every 16S sequence matched with Pseudomonas sp\n",
    "\n",
    "Why is N50 is useful statistic to calculate? Why not just list the mean or median contig length? In your assembly, are scaffolds longer than contigs, or are scaffolds approximately equal to contigs?\n",
    "A way to think about N50 is that >= half of the nucleotides in the assembly belong to contigs with length >= N50\n",
    "\n",
    "This is much better than the mean or median contig length because it excludes very short contig lengths which are equally important in the assembly.\n",
    "\n",
    "Scaffold length is longer in our contigs very marginally.\n",
    "\n",
    "### Research and write-up\n",
    "\n",
    "Antibiotic resistant bacteria has become a problem that is rapidly growing, particularly multidrug resistant bacteria (resistant to more than two drugs). From 1997-2002 the percentage of multidrug-resistant strains increased from 13-22%. Pseudomonas is a genus of bacteria that is able to rapidly mutate through the time frame of treatment for antibiotics. It is extremely resilient and survives in areas such as swimming pools, contact lens solution, humidifiers, soil and vegetables. It is also thrives in hospitals, including respiratory equipment, antiseptics, soap, sinks, mops, medicine, and hydrotherapy pools.\n",
    "\n",
    "According to our KEGG map for our sequence when using the RAST, 100% of Pseudomonas species are beta-Lactam resistant, which includes resistance to antibiotics such as penicillin. There are multiple pathways for beta-Lactam resistance to be found, including pathways caused by Cephamycin C and Cephalosporin C. There are 14 distinct ECs that control the biosynthesis of penicillin and cephalosporin, 4 of which are found in Pseudomonas. We also observed 4 distinct ECs for biosynthesis of Tetracycline, indicating the possibilities of resistance to Tetracycline antibiotics. This is consistent with a paper relating the functionality of the efflux pump to be causal for antibiotic resistance. The efflux pump is a transmembrane protein that allows amphiphilic drugs into the cell, which when mutated, nullifies a variety of antibiotics such as fluoroquinolones, tetracycline, chloramphenicol, macrolides.\n",
    "\n",
    "The genome of Pseudomonas has multiple antibiotic resistant genes that influence multiple pathways, making effective treatment for complex. Treatment for this bacteria can be complex because the rapid rate of mutation can easily exceed the time frame for antibiotic treatment for the patient. Pseudomonas also is able to thrive and is more likely to become resistant in immunocompromised patients, making this bacteria extremely dangerous. The strategy for treatment is generally to give multiple types of antibiotics, however this poses a pressing concern for evolutionary strains that are multidrug resistant.\n",
    "\n",
    "Source: Lister, Philip D., et al. “Antibacterial-Resistant Pseudomonas Aeruginosa: Clinical Impact and Complex Regulation of Chromosomally Encoded Resistance Mechanisms.” Clinical Microbiology Reviews, American Society for Microbiology (ASM), Oct. 2009, www.ncbi.nlm.nih.gov/pmc/articles/PMC2772362/#r234."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
